Background image: add support for absolute theme path URLs

ramonjd opened the pull request:
Moved to https://github.com/WordPress/gutenberg/pull/61401

> [!NOTE]
> This PR is experimental and will be a reference for ongoing work. See https://github.com/WordPress/gutenberg/pull/60578#issuecomment-2087925879

## What? And How??!!!

Part of:

- https://github.com/WordPress/gutenberg/issues/54336

This PR proposes a way to add support for theme relative url paths for theme.json/templates/patterns by:

- for the frontend, checking for `"source": "theme"` on the background image object in background supports, and appending the current theme path. It uses `get_stylesheet_directory_uri` instead of `get_template_directory_uri` so that the path will reflect any active child theme.
- for the editor, adding a new property to editor settings `currentTheme`, that contains the current theme directory path. It also uses `get_stylesheet_directory_uri` instead of `get_template_directory_uri` so that the path will reflect any active child theme.


## Why?

So that themes don't have to use global styles and rather, insert theme-relative paths into templates, patterns and theme.json.

This enhances portability of themes as Gutenberg will use the theme's current directory, regardless of install.

## Testing Instructions

Assuming that `img/Untitled.jpg` exists in the theme dir, here's some test theme.json. Note only top-level background images are possible until https://github.com/WordPress/gutenberg/pull/60100 gets in.

```json
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"settings": {
		"appearanceTools": true
	},
	"styles": {
		"background": {
			"backgroundImage": {
				"source": "theme",
				"url": "img/Untitled.jpg"
			},
			"backgroundPosition": "center center",
			"backgroundRepeat": "no-repeat",
			"backgroundSize": "cover"
		}
	}
}
```

Here's a test pattern (could also be HTML). Add it under `/patterns` in the theme directory:

```php
<?php
/**
 * Title: Empty theme pattern
 * Slug: emptytheme/empty-pattern
 * Categories: featured
 */
?>
<!-- wp:group {"style":{"background":{"backgroundImage":{"url":"img/Untitled.jpg","source":"theme"}},"dimensions":{"minHeight":"299px"}},"layout":{"type":"constrained"}} -->
<div class="wp-block-group" style="min-height:299px"></div>
<!-- /wp:group -->
```

It should also work with templates. Add it under `/templates` in the theme directory:

```html
<!-- wp:group {"style":{"background":{"backgroundImage":{"url":"img/Untitled.jpg","source":"theme"}},"dimensions":{"minHeight":"500px"}},"layout":{"type":"constrained"}} -->
<div class="wp-block-group" style="min-height:500px">
<!-- wp:template-part {"slug":"header","tagName":"header","theme":"emptytheme"} /-->
<!-- wp:query {"queryId":1,"query":{"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","sticky":""}} -->
<div class="wp-block-query">
	<!-- wp:post-template -->
	<!-- wp:post-title {"isLink":true} /-->
	<!-- wp:post-excerpt /-->
	<!-- /wp:post-template -->
</div>
<!-- /wp:query -->
</div>
<!-- /wp:group -->
```

1. For each approach above, check that the background image's absolute path is rendered into the `background-image` CSS value for both the frontend and editor.
2. Ensure to check both the site editor (global styles) and post editor (block supports).

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

ramonjd committed:
Supporting background.backgroundImage.source === "theme" relative paths

ramonjd committed:
This is an experimental commit to test out how to:
- get the theme directory URI, and
- ensure background styles are parsed for theme.json and pattern/template files

ramonjd commented on lib/block-supports/background.php:
```diff
@@ -30,6 +30,27 @@ function gutenberg_register_background_support( $block_type ) {
+	/*
+	 * "theme" source implies relative path to the theme directory
+	 */
+	if ( ! empty( $background_styles['backgroundImage']['url'] ) && 'theme' === $background_image_source ) {
```
This will take care of rendering block and global styles on the frontend.

ramonjd commented on lib/compat/wordpress-6.6/rest-api.php:
```diff
@@ -29,3 +29,36 @@ function wp_api_template_access_controller( $args, $post_type ) {
+ * Registers additional fields for WP_REST_Themes_Controller class.
+ * https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php
+ */
+function gutenberg_register_wp_theme_directory_uri_field() {
```
This is the test. 

Is there another way to get the theme directory to the editor?

ramonjd commented on packages/editor/src/components/provider/use-block-editor-settings.js:
```diff
@@ -116,19 +117,24 @@ function useBlockEditorSettings( settings, postType, postId ) {
 					'allowRightClickOverrides'
 				),
 				blockTypes: getBlockTypes(),
+				currentTheme: {
```
Here's where my lack of imagination or sleep failed me. I couldn't think of another way, though there probably is one, to ensure the current theme data is available in the editor context.

ramonjd commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -369,6 +388,12 @@ const elementTypes = [
 ];
 
 function useBlockProps( { name, style } ) {
+	const { themeDirURI } = useSelect( ( select ) => {
```
Not sure about this either.

For the record, this PR should just take care of theme.json/global styles but it would be good to know in advance how we'll deal with blocks.

These blocks could come from template or pattern files as well.

andrewserong reviewed (commented):
Thanks for digging in! Hope you don't mind, but looking over the code change sparked a couple of thoughts, so I thought I'd jot them down here in case it's helpful 🧠 📔

> Is there another way to get the theme directory to the editor?

Great question. Also, to further complicate matters, in PHP the `get_theme_file_uri()` function will resolve files from a child theme (i.e. checks if the file exists in the child theme, and if not, moves up to the parent theme to see if the file exists, and then it also supports the `theme_file_uri` filter). This presents some challenges for the editor as we can't call `get_theme_file_uri()` with each individual file that might be in a template or `theme.json`.

A couple of off-the-top-of-my-head thoughts:

* If we don't support resolving between child and parent themes, then if someone goes to create a child theme from a theme that includes background images in `theme.json` or patterns, then the background images will likely be broken in the child theme.
* One potential workaround could be to also output `get_template_directory_uri()` in the theme API endpoint?
* Or, a more complex but potentially explicit approach could be: what if it were possible to register or declare theme assets that are available within a theme? I.e. either in `theme.json` and/or via a PHP function? Then we could have an endpoint that returns the URLs for each of the registered assets.

That last point could be a "good" way to handle it explicitly, but even as I write that, I can't imagine it being at all simple to come up with a reliable way to do it that everyone would agree to. So, another potential idea:

What if instead of extending the `theme` API endpoint, we injected the theme URI (and potentially the template / parent theme URI) into block editor settings via the `'block_editor_settings_all'` hook? That way we'd be hacking the URIs into block settings without touching any API endpoints?

ramonjd commented:
Thanks for looking at this one @andrewserong! 🙇🏻 

> Also, to further complicate matters, in PHP the get_theme_file_uri() function will resolve files from a child theme (i.e. checks if the file exists in the child theme, and if not, moves up to the parent theme to see if the file exists, and then it also supports the theme_file_uri filter)

Yeah - I used `get_stylesheet_directory_uri` in the endpoint to test out some things as well. That or  `get_template_directory_uri()` should work.

> What if instead of extending the theme API endpoint, we injected the theme URI (and potentially the template / parent theme URI) into block editor settings via the 'block_editor_settings_all' hook? That way we'd be hacking the URIs into block settings without touching any API endpoints?

Definitely more pragmatic, given we're touching settings anyway. I like!

> Or, a more complex but potentially explicit approach could be: what if it were possible to register or declare theme assets that are available within a theme?  

This would be more bullet proof, but I agree that it might be a larger feature that would require wider buy in. Also it introduces another registration layer for folks to think about. 

An MVP could be `get_stylesheet_directory_uri() || get_template_directory_uri() + $path_to_image` for now. What do you reckon?

andrewserong commented:
> An MVP could be get_stylesheet_directory_uri() || get_template_directory_uri() + $path_to_image for now. What do you reckon?

Sounds worth trying to me! Then, once things are working well enough in both the editor and server-side, we could ping Gutenberg core folks for more visibility to confidence check the approach?

ramonjd committed:
Roll back API changes and insert into editor settings under `currentTheme` both parent and child dir
Update tests to check for image paths with leading slash

Revert rogue tab

ramonjd commented on lib/block-editor-settings.php:
```diff
@@ -76,6 +76,13 @@ function gutenberg_get_block_editor_settings( $settings ) {
 	}
 
+	// Could house also `__unstableIsBlockBasedTheme` as 'isBlockTheme' => $current_theme->is_block_theme().
+	$current_theme            = wp_get_theme();
```
Using `wp_get_theme` to get full `WP_Theme` object in case folks want to use [other helpful methods](https://developer.wordpress.org/reference/classes/wp_theme/).

ramonjd committed:
If we pass a 'baseUrl' to the style engine, it will it append it to all url-like paths.

Revert rogue space

ramonjd commented on packages/style-engine/class-wp-style-engine.php:
```diff
@@ -605,9 +610,15 @@ protected static function get_url_or_value_css_declaration( $style_value, $style
-					$value = "url('" . $style_value['url'] . "')";
+				if ( ! empty( $style_value['url'] ) && is_string( $style_value['url'] ) ) {
+					$url = $style_value['url'];
+					if ( ! empty( $options['base_url'] ) ) {
```
So the idea here is that if you pass a base url to the style engine it will append it to all URL-like values.

Trying to work out if this is a bad idea and whether the url should come already edited.

Pros:

- A quick way to create absolute URLs for multiple values
- Centralized


Cons:

- It shouldn't interfere with other URLs, e.g., `url(data:image/png;base64,iRxVB0…)`, `url(myFont.woff)`, `url(#IDofSVGpath)`; Maybe we could also check for `$style_value['source'] === "theme"` to get around that.

andrewserong replied:
Neat idea to centralize! Some quick thoughts off the top of my head:

* If we're unsure of the long-term shape of anything, place the logic in the consumer of the style engine, rather than in the style engine itself. (I.e. in this case, in `background.php`).
* If we're sure of the shape, then centralizing is a good idea. In this case, I'd lean toward calling it something like `'theme_base_url'` and only prepending if `$style_value['source'] === "theme"` as you mention.

ramonjd replied:
Thanks! I was about to ping you 😄 

> If we're unsure of the long-term shape of anything, place the logic in the consumer of the style engine, rather than in the style engine itself. (I.e. in this case, in background.php).

That's a great summary.

Are we ever sure of anything? 😄 

This was mainly an experiment to see if reduced the code (a bit) and worked (yes).

Considering how new background supports is, I might err on the side of caution and roll back to what we had. It's not pretty, but it's more flexible.

andrewserong replied:
> Are we ever sure of anything? 😄

In Gutenberg? Great question 😆

> Considering how new background supports is, I might err on the side of caution and roll back to what we had. It's not pretty, but it's more flexible.

Sounds good! And you've got a solid plan for a follow-up PR if we decide we'd like to centralise 👍

ramonjd commented on packages/style-engine/src/styles/background/index.ts:
```diff
@@ -8,19 +8,21 @@ const backgroundImage = {
-			_backgroundImage?.url
-		) {
+		if ( typeof _backgroundImage === 'object' && _backgroundImage?.url ) {
+			const url = !! options?.baseUrl
```
So this was mainly so we wouldn't have to add any ugly code to `useBlockProps`. 

In a previous commit I was trying out modifying the style attribute object:

https://github.com/WordPress/gutenberg/pull/60578/commits/4a0d1075ececa5237c164ed6fe36a2ec31af3f80#diff-292053c9ec98ebdd73943767587e8093315ad3a2bf01ca3b5352efc48326becdR322

ramonjd committed:
Rolls back f7bf075d4a0b143008f1809c8254ebab61492e00

ramonjd committed:
Removed useBlockProps and now editing style object directly

ramonjd commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -391,6 +394,18 @@ export function getStylesDeclarations(
+	 * Only applies to background styles for now.
+	 */
+	if ( !! blockStyles?.background ) {
+		blockStyles = setBackgroundStyleDefaults( blockStyles, {
```
Will also cater for https://github.com/WordPress/gutenberg/pull/60405 and https://github.com/WordPress/gutenberg/pull/60100

ramonjd committed:
Remove unused "parentDirectoryURI" prop in block editor settings
Added/updated unit tests

Bad horsie

ramonjd commented on packages/style-engine/src/styles/background/index.ts:
```diff
@@ -8,11 +8,7 @@ const backgroundImage = {
-			_backgroundImage?.source === 'file' &&
-			_backgroundImage?.url
-		) {
+		if ( typeof _backgroundImage === 'object' && _backgroundImage?.url ) {
```
Removing `_backgroundImage?.source === 'file'` check as we only care if background image is an object. There's no special processing for `file` or `theme` in the style engine.

ramonjd committed:
Yes, spacing.

ramonjd committed:
Unit test no styles

ramonjd commented on packages/block-editor/src/hooks/style.js:
```diff
@@ -312,6 +319,15 @@ export function addSaveProps(
+	 * Only applies to background styles for now.
+	 */
+	if ( !! style.background ) {
+		style = setBackgroundStyleDefaults( style, editorSettings );
```
The rationale is that some style values need to have sensible defaults that are output as CSS, but not necessarily part of the styles attribute. 

`setBackgroundStyleDefaults` could one day be `setStylesDefaults` or something.

This could theoretically be done in the style engine if we wanted to make the style engine opinionated.

noisysocks commented:
Is adding `stylesheet_uri` and `template_uri` to the `/v2/themes` in the REST API an option? That way we can use the existing `wp.data.select( 'core' ).getActiveTheme()`.

In my mind Gutenberg is a REST API application. Part of the original motivation was to stress test the REST API, so if something is missing from the REST API then it usually implies that we should add it. Editor settings should be primarily for configuring the default state of the editor, not passing data from server to client.

ramonjd commented:
> Is adding stylesheet_uri and template_uri to the /v2/themes in the REST API an option? That way we can use the existing wp.data.select( 'core' ).getActiveTheme().

Thanks @noisysocks Answer: Yes 😄  

I went in that exact direction in the first iteration of this PR

https://github.com/WordPress/gutenberg/commit/99c6fd6fa0bbcd9bd0ac0f0fede9705630cd1e3d

Ultimately it was less complicated to add the values to the settings directly (I was adding them to the editor settings anyway to expose it to the editor(s)). 

See for context https://github.com/WordPress/gutenberg/pull/60578#pullrequestreview-1988111048

I still think the API route has legs, but another issue I was facing (if I remember correctly) was using the core data package `@wordpress/core-data` in the block editor package, and I was adding it to the editor settings anyway to expose it to the editor(s). 

The theme path is required wherever Gutenberg is compiling global and block styles, so post and site editors.

ramonjd commented:
> The theme path is required wherever Gutenberg is compiling global and block styles, so post and site editors.

I suppose the results of the API call could be mashed into [useSpecificEditorSettings](https://github.com/WordPress/gutenberg/blob/5eee07e881aa6bfb9837323e097a0a3996deff4f/packages/edit-site/src/components/block-editor/use-site-editor-settings.js#L115-L115) for the site editor and whatever the equivalent is for the post editor. 

It's still injecting theme config into editor settings but through the back door 🤷🏻

noisysocks commented:
> I still think the API route has legs, but another issue I was facing (if I remember correctly) was using the core data package `@wordpress/core-data` in the block editor package, and I was adding it to the editor settings anyway to expose it to the editor(s).

Right, yes, that's a good observation. I think this pokes at the heart of the problem with how we're architecting things here which is that _themes_ are a WordPress concept, `@wordpress/block-editor` can be used outside of WordPress, and so we're trying to add a feature to `@wordpress/block-editor` that relies on the concept of _themes_ (`source: 'theme'`, `themeDirURI`, etc) to something that doesn't (shouldn't) know what a theme is.

Is my understanding correct? A lot of this is unfamiliar to me 😀

Do we use `source: 'theme'` anywhere else? If so perhaps there's an existing approach we source inspiration from.

One idea I have is that maybe `@wordpress/editor` (which _does_ know what a theme is) can rewrite the styles object from `"backgroundImage": { "url": "/assets/images/my_image.png", "source": "theme"}` to `"backgroundImage": { "url": "https://example.com/wp-content/themes/mytheme/assets/images/my_image.png"}` before handing it to `@wordpress/block-editor`.

andrewserong commented:
> Do we use source: 'theme' anywhere else? If so perhaps there's an existing approach we source inspiration from.

To my knowledge, no. Part of the goal with this feature is to create a good example for this that could then potentially be used for other features, too. So getting the architecture for it right will likely be a part of the process here, I think.

One thing to consider re: including the prefix in the API endpoint is the balance between pragmatically injecting something into block editor settings vs extending API responses. I imagine the latter might be a bigger discussion as it could have implications beyond what's required for the feature to work within the editor? I suppose either way, though, it's making changes that'll need to be maintained, so it's good to think through each of the options carefully.

noisysocks commented:
> I imagine the latter might be a bigger discussion as it could have implications beyond what's required for the feature to work within the editor?

Yes but in this case I think it seems like an omission that there's no way to get the base URL of a theme via the themes endpoint. I'd caution against viewing the REST API as an immaculate piece of software that we can't touch. It's a relatively new part of WordPress's codebase and a lot of its design has been in fact driven by Gutenberg's needs.

andrewserong commented:
> Yes but in this case I think it seems like an omission that there's no way to get the base URL of a theme via the themes endpoint.

Very good point! I've opened [a Trac ticket](https://core.trac.wordpress.org/ticket/61021) so we can start the discussion and explore proposing it in core 👍

https://core.trac.wordpress.org/ticket/61021

andrewserong commented:
I've thrown up a quick core PR for the API endpoint idea, too, to see if that helps the discussion as well: https://github.com/WordPress/wordpress-develop/pull/6399

ramonjd commented:
@andrewserong @noisysocks Thanks for the great discussion, and helping me move this PR forward. 🙇🏻  

> Yes but in this case I think it seems like an omission that there's no way to get the base URL of a theme via the themes endpoint.

👍🏻 

> I am thinking ideally @wordpress/editor (which knows what a theme is) should handle translating theme-aware URLs into regular URLs that @wordpress/block-editor can understand. So for example @wordpress/editor can rewrite the styles object from "backgroundImage": { "url": "/assets/images/my_image.png", "source": "theme"} to "backgroundImage": { "url": "https://example.com/wp-content/themes/mytheme/assets/images/my_image.png"} before handing it to @wordpress/block-editor.

This is the part that needs investigation.

Without looking at the code (yet), the editor package can call `@wordpress/core-data`. Whatever mutated properties or transformative functions it provides still need to be available to the block editor package.

`useBlockEditorSettings` is how this PR was doing it in the first place, using the return value from the `/themes` endpoint.

One thing I'd like to try today is to transform any style objects that pass through `useBlockEditorSettings` on their way to the block editor. 👍🏻 

> I've thrown up a quick core PR for the API endpoint idea, too, to see if that helps the discussion as well: https://github.com/WordPress/wordpress-develop/pull/6399

Noice! I think it's definitely worth a Core PR - thanks for getting the conversation going early. It's where I was trying to go with https://github.com/WordPress/gutenberg/pull/60578/commits/6fc2a0a2e7ee7b17cfe8ca6833d6215c4e226f9c#diff-ac1b5f0227617fa339610c02e5f8f1c191bbdcc3c0735d1b8bb45637725ea861R50 so not a huge paradigm shift.

ramonjd commented:
> One thing I'd like to try today is to transform any style objects that pass through useBlockEditorSettings on their way to the block editor. 👍🏻

Okay, this was a fanciful notion 😄 

Gutenberg doesn't (yet) pass theme style objects to the editor, only the rendered styles themselves. There are two PRs that do this, albeit in different ways.

- https://github.com/WordPress/gutenberg/pull/59929
- https://github.com/WordPress/gutenberg/pull/57908 ([here](https://github.com/WordPress/gutenberg/pull/57908/files#diff-194fa91cbbc3043754b8f48d4ab6964bf37abdef44d11b07b24c92d8a651833aR81))

The latter seems like it will be merged before the former. 

For this reason I can't modify the styles before they get to the editor. Furthermore, the same would have to be done for block styles.

I think the original approach in this PR, to pass down theme URIs in settings via the editor provider, is the most accessible.

The settings already contain many WordPress/CMS/Theme specific values, e.g., `siteUrl `, `maxUploadFileSize`, `__experimentalFeatures` (theme.json settings), `__unstableIsBlockBasedTheme`....

I agree on the separation of concerns. The provider in 

Options:

- Add the current theme URI to the settings, maybe underneath `__experimentalTheme = {}` or something
- Do the above, but edit the settings object in `edit-post` and `edit-site`. Same same.
- Do an `apiFetch( { path: `/wp/v2/themes` } )` in `block-editor` somewhere, in a private action or something. But then we'd need the current theme anyway. 👎🏻 
-  ??

ramonjd commented:
> Add the current theme URI to the settings, maybe underneath __experimentalTheme = {} or something

I'll do this for now, to get things working. Basically reverting back to original commit.

I'd be happy to continue the discussion here, and change things as they evolves.

> Add a new method to the settings (like onNavigateToEntityRecord does it) that returns theme/WP/consumer-related info.

It makes me think: do we need to think about how to separate block editor and consumer/WordPress etc concerns somehow in these settings? 

Maybe the block editor could have access to two selectors:

- a `getSettings()`, which strictly returns editor settings, and then 
- a `getContextSettings()` (need better name) that returns environmental settings, e.g., theme, whatever

I don't know if that's a good idea, and it'd be a big refactor. Nevertheless, we could standardize some properties related to the consumer so block editor instances outside WordPress could populate this stuff too.

```js
{
    assetsUri,
    siteHomePage,
}
```

Related issue:

- https://github.com/WordPress/gutenberg/issues/59536

ramonjd committed:
Getting URI from themes API response

ramonjd committed:
Switching to experimental

ramonjd commented on lib/compat/wordpress-6.6/rest-api.php:
```diff
@@ -29,3 +29,76 @@ function wp_api_template_access_controller( $args, $post_type ) {
+	 * Adds `stylesheet_uri` fields to WP_REST_Themes_Controller class.
+	 * Core ticket: https://core.trac.wordpress.org/ticket/61021
+	 */
+	function gutenberg_register_wp_rest_themes_stylesheet_directory_uri_field() {
```
All this needs to match https://github.com/WordPress/wordpress-develop/pull/6399

ramonjd commented:
Pinging @WordPress/gutenberg-core for wider opinions. 

## TL;DR

The block editor receives "environment" settings via `getSettings()` e.g., `siteUrl `, `maxUploadFileSize`, `__experimentalFeatures` (theme.json settings), `__unstableIsBlockBasedTheme` among others. That is, values that are not strictly editor-specific, but rather site or WordPress (or consumer) settings.

This PR proposes to introduce another one with current theme properties. The main question is should we, and, if so, how?

Is there an acceptable (or alternative way) to access core data or other such data in the block editor without overloading the settings objects with such "environment" properties? I think question has come up many times ([example](https://github.com/WordPress/gutenberg/pull/59929#discussion_r1531747404)), and it will again. 

I'm not really sure how, but would it help to split out or standardize the interface for "site" settings?

All advice and input welcome 😄

ramonjd commented on packages/editor/src/components/provider/use-block-editor-settings.js:
```diff
@@ -290,6 +294,9 @@ function useBlockEditorSettings( settings, postType, postId, renderingMode ) {
 			// Check these two properties: they were not present in the site editor.
 			__experimentalCreatePageEntity: createPageEntity,
 			__experimentalUserCanCreatePages: userCanCreatePages,
+			__experimentalCurrentTheme: {
```
This could be rather `theme: {}` or `siteTheme` or :shrug:

Mamaduka commented:
> I'm not really sure how, but would it help to split out or standardize the interface for "site" settings?

We already have the site settings - `wp.data.select( 'core' ).getSite()`, but the `/settings` endpoint is pretty strict about user capabilities. Should anyone who can access the editor have access to this setting?

You can also make the setting private if you're unsure that it will stay there. Here's one example - #58239.

ramonjd commented:
Thanks @Mamaduka!! 

> Should anyone who can access the editor have access to this setting?

For the one proposed in this PR, yes. It's the stylesheet directory URI value.

> You can also make the setting private if you're unsure that it will stay there.

Yeah, that is an option I was looking at too, thanks! I think I was just trying allay any concerns about overloading the settings, but I can't see any other way to get the data to the editor.

ramonjd commented:
> "before I had my coffee" ideas

I've just had a coffee, and I'm not I'm any better for it 😄 

> I just remembered that classic themes supported the background image; see the 2016 theme. I think we can use the same storage method here and maybe expose the value as part of the active theme schema.

I looked into how WordPress currently supports background image properties via [get](https://github.com/WordPress/wordpress-develop/blob/09254045ed41e82c05637f42f2ca6d188b11747a/src/wp-includes/theme.php#L1898) and [set_theme_mod](https://github.com/WordPress/wordpress-develop/blob/09254045ed41e82c05637f42f2ca6d188b11747a/src/wp-admin/includes/class-custom-background.php#L634) and printed out in a [custom style sheet](https://github.com/WordPress/wordpress-develop/blob/09254045ed41e82c05637f42f2ca6d188b11747a/src/wp-includes/theme.php#L1945).

Themes can specific a body background image like this:

```php
add_theme_support(
    'custom-background',
    array(
        // As well as default image, there's default-position-x, default-position-y, default-size, default-repeat, default-attachment
        'default-image' => get_stylesheet_directory_uri() . '/img/bird.jpg',
    )
);
```

Or like this:

```php
add_theme_support( 'custom-background' );
// As well as background_image, there's background_preset, background_size, background_position_x, background_position_y
set_theme_mod( 'background_image',  get_stylesheet_directory_uri() . '/img/bird.jpg' );
```

Setting background images via theme mods will overwrite any theme.json or global styles images set in the site editor.

<img width="644" alt="Screenshot 2024-04-22 at 11 50 44 am" src="https://github.com/WordPress/gutenberg/assets/6458278/25716cab-b83b-4036-8b57-c2605d76bc9f">


❓ **Just thinking out loud:** should Gutenberg check for the presence of a theme mod background image and associated properties, and integrate it with global styles? The next, related question is, should setting background images in the site editor overwrite customizer/theme_mod backgrounds? I think that's a separate issue to this PR, I'll add a note to the [tracking issue](https://github.com/WordPress/gutenberg/issues/54336).

This PR is about allowing themes to bundle assets and maintain theme portability. A theme developer needs only to specify a relative path to the background image in theme.json and elsewhere, and it'll work in every installation of the theme.

> I'm not sure if REST API should expose directory paths. Usually, all media is stored in uploads, and then REST API returns absolute paths.

This PR is really only considering theme assets, and not images or other media that are uploaded to a WordPress site before or after theme installation. Portability is a main concern here.

ramonjd commented:
> Just thinking out loud: should Gutenberg check for the presence of a theme mod background image and associated properties, and integrate it with global styles? The next, related question is, should setting background images in the site editor overwrite customizer/theme_mod backgrounds? I think that's a separate issue to this PR, I'll add a note to the https://github.com/WordPress/gutenberg/issues/54336.

Added new issue here:

- https://github.com/WordPress/gutenberg/issues/60939

ramonjd commented on lib/block-supports/background.php:
```diff
@@ -30,6 +30,30 @@ function gutenberg_register_background_support( $block_type ) {
+		if ( ! str_starts_with( $background_styles['backgroundImage']['url'], '/' ) ) {
+			$background_styles['backgroundImage']['url'] = '/' . $background_styles['backgroundImage']['url'];
+		}
+		$background_styles['backgroundImage']['url'] = esc_url( get_stylesheet_directory_uri() . $background_styles['backgroundImage']['url'] );
```
I think we could safely use [get_theme_file_uri](https://developer.wordpress.org/reference/functions/get_theme_file_uri/) here. It'll look in the active theme dir first (e.g., active child theme), and then look to the parent theme directory if not found.

Why? In the case where a child theme is active, I can imagine that some themes might want to dictate that relative image paths should be prefixed with a theme parent directory URI, and not the active child theme's.

This could also be expressed explicitly in theme.json later if we come up with a decent structure/naming

```json
	"styles": {
		"background": {
			"backgroundImage": {
				"source": "theme-child", // or "theme-parent" if active is default, or "theme-inherit" or 🤷🏻 
				"url": "img/Untitled.jpg"
			},
		}
	}
```

ramonjd replied:
The above comment doesn't consider the editor or background images in template or pattern files 🤔 

So perhaps we go the theme.json explicit route. It would be more flexible for theme developers anyway, in the sense that they could mix and match where images came from, e.g., site background image from the parent theme and other from child theme.

andrewserong replied:
> I think we could safely use [get_theme_file_uri](https://developer.wordpress.org/reference/functions/get_theme_file_uri/) here.

Good catch! Yes, that logic looks good for allowing child themes to override an image from a parent theme, with fallback to the parent theme when the child theme doesn't override it 👍

> So perhaps we go the theme.json explicit route. It would be more flexible for theme developers anyway, in the sense that they could mix and match where images came from, e.g., site background image from the parent theme and other from child theme.

That sounds like a good potential idea for a follow-up to me. Allowing child themes to universally override something from a parent theme conforms to the general principle of child theming, and I'm curious if many folks would intentionally want to go with the exception? (Explicitly using `theme-parent` to skip an overridden image in a child theme feels potentially a bit edge-casey to me).

What blockers would we encounter if we go with `"source": "theme"` for now and attempt to have the fallback logic be: use the child theme asset if available, and fallback to the parent theme if not?

ramonjd replied:
Thanks for chiming in! I'm probably overthinking it. 

I'm only raising this now as this PR introduces `"source": "theme"` so I wanted to consider this angle while we have the chance to redefine it.

> What blockers would we encounter if we go with "source": "theme" for now and attempt to have the fallback logic be: use the child theme asset if available, and fallback to the parent theme if not?

This would be ideal, and I think the way forward. Especially seeing a child theme inherits templates, theme. json etc from the parent theme.

The only splinter is reproducing the file checking in the editor, which could be a performance hit assuming we set up some API. 

🤔

ramonjd replied:
> The only splinter is reproducing the file checking in the editor, which could be a performance hit assuming we set up some API.

Unless we come up with a way to reliably parse template/pattern files.

This PR as it is should be okay just for theme.json, but I'm concerned that if a child theme is using its parents' templates/patterns the file paths need to be resolved with the parent's URI and not the child's.

It's all very spicy!

andrewserong replied:
> I'm only raising this now as this PR introduces "source": "theme" so I wanted to consider this angle while we have the chance to redefine it.

Good idea, this is the right time to be thinking through all the options, for sure 👍

> The only splinter is reproducing the file checking in the editor, which could be a performance hit assuming we set up some API.

Yeah, the lightest weight version I could think of would be to do a `HEAD` request using the child theme url, and if we get a 404, then use the parent them url, or something like that?

ramonjd replied:
> Yeah, the lightest weight version I could think of would be to do a HEAD request using the child theme url, and if we get a 404, then use the parent them url, or something like that?

I've been investigating this, and there's a slight issue in that it's an async call that needs to be made in a series of synchronous steps, either when parsing the styles or the block props. 

There's no clear entry point for this and I'm a bit stumped as to where the relative URL parsing should therefore take place without impacting performance.

Parsing the every theme json style either in the backend or frontend might work, but it's intensive.

I think I've come up with an interim workaround. For now, in the editor, I'm testing doubling up the images, that is, adding both urls to the `background-image` property with priority given to the `stylesheetURI`, e.g., 


```css
background-image: url("http://localhost:8888/wp-content/themes/emptytheme-child/img/untitled.png"), url("http://localhost:8888/wp-content/themes/emptytheme/img/untitled.png");
```

Where the first image doesn't load, the second image is displayed.

This works of course, and can even be combined with linear gradients, 

```css
background-image: linear-gradient(rgba(0, 0, 255, 0.5), rgba(255, 255, 0, 0.5)), url("http://localhost:8888/wp-content/themes/emptytheme-child/img/untitled.png"), url("http://localhost:8888/wp-content/themes/emptytheme/img/untitled.png");
```


but I hope it doesn't bite us in the derrière later when we deal with more complicated layers. I

 can't see how yet though. Even if we implement layers as described in https://github.com/WordPress/gutenberg/pull/60739#issuecomment-2081872934, we could duplicate the properties for each image layer in the editor. 🤷🏻 

Anyway, that's enough for me today.

andrewserong replied:
> For now, in the editor, I'm testing doubling up the images, that is, adding both urls to the background-image property with priority given to the stylesheetURI, e.g.

Oh, that's pretty clever! The main edge case I can think of is if a theme includes a transparent image like a parchment texture that's meant to be combined with a background color, and if a child theme wanted to override that with a different transparent texture. If we add the two together, it sounds like that would then layer the two images, instead of visually overriding them?

ramonjd replied:
> The main edge case I can think of is if a theme includes a transparent image like a parchment texture that's meant to be combined with a background color, and if a child theme wanted to override that with a different transparent texture. If we add the two together, it sounds like that would then layer the two images, instead of visually overriding them?

Yeah, that sounds like a risk. Thanks for raising it. 

I'm tending towards accepting it as a reasonable risk in order to get the feature working, at least while we come up with a better approach.

I'm also going to keep exploring - something might drop into my lap 😅

ramonjd committed:
Testing get_theme_file_uri on the backend and duping the background image on the frontend.

draganescu commented:
👋🏻 I haven't got reading all the code but my 1st question is what other sources are?

ramonjd commented:
> I haven't got reading all the code but my 1st question is what other sources are?

Thanks for the question @draganescu (And the code is a bit hacky at the moment. 😅 I'm going to clean it up and update tests today.)

Thus far, sources considered are:

- Absolute paths (currently supported), so remote file resources
- Theme resource accessible via `get_theme_file_uri`. When the `source` key is set to `"theme"`, the implication here is that it's a file in the active child or parent theme.
- The plan is to also have `source` support featured images, but I'm thinking this would initially be a editor feature.

The primary goal of this PR is to allow background image relative paths to theme resources for the following entities:

- theme.json
- template/parts
- patterns

@andrewserong also raised the scenario of previewing themes, whereby we'd specify other theme URIs, potentially via filters, so that the relative paths resolve to installed theme dirs. That's a later task 😄 

I think that's the "is" and "will be" state roughly described 😄 

What do you think? Should Gutenberg consider more internal/external sources? Thanks for any tips you have!

draganescu commented:
Thank you for working on this intricate issue! 

> via the "source": "theme" flag

The thing is this is not a flag, it's a prop we introduce to the theme.json schema and the question is if this is superfluous or not. if we only really have one source its not great. Because if we really only have one source, can't we do basic string searches to see if the URLs are not absolute and then default to theme relative if they're not?

ramonjd commented:
> The thing is this is not a flag, it's a prop we introduce to the theme.json schema and the question is if this is superfluous or not. 

Prop/flag/horse/spaghetti - you get where I'm going 😆 

Answering the [original question](https://github.com/WordPress/gutenberg/pull/60578#issuecomment-2082460741), insofar as theme.json/template/pattern files are concerned there are at least two sources: a parent/child theme directory, and any URL.

As to whether it's superfluous, in relation to theme.json/template/pattern files, I can see an argument for the proposition.

There was talk of other sources  "featured-image" (only works for posts with support for featured images), and also image  sites like OpenVerse, but I imagine that'd be confined to the editor. 

So it might be handy in the block editor context, but perhaps superfluous as you suggest, for theme assets.

> can't we do basic string searches to see if the URLs are not absolute and then default to theme relative if they're not?

Yes, that's definitely option, thanks for the idea! It does seem a bit redundant if we can't think of any other sources for theme.json/template/pattern files. 

I think @andrewserong had reasons for introducing it, so I'll defer to his input before removing it in this PR and from the theme.json schema. It's not being used in the background to any great effect.

andrewserong commented:
> There was talk of other sources "featured-image" (only works for posts with support for featured images)

Yes, the original idea was for three intended states (full urls, theme relative urls, and featured images) when we were looking at the shape of this attribute at the block level in post content. So a Group block could be set to any of those three, with the potential to have other arbitrary sources further down the line as you mention, but I don't think there were any concrete ideas about that.

Personally I like the idea of the explicit prop for the source, as it feels more direct than determining based on the presence of `http` at the beginning of the string. And if we remove `theme` as an explicit prop, we could lose the ability to do other things in the future, such as if the relative path should actually refer to an uploads directory? That's just an idea though, but I'd be a little cautious about making it implicit.

ramonjd committed:
Just a commit to explore adding a private selector to mimic get_theme_file_uri (would be extracted into a new PR)
Haven't worked out global styles yet. CSS rules might need to be injected into the DOM???

ramonjd committed:
whoops

ramonjd commented on packages/block-editor/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -1230,6 +1246,7 @@ export function useGlobalStylesOutputWithConfig( mergedConfig = {} ) {
 			disableLayoutStyles,
 			isTemplate
 		);
+		console.log( 'globalStyles', globalStyles );
```
This is a marker to remind me that `getThemeFileURI` isn't implemented for global styles yet either.

Maybe it needs to be done by injecting a CSS rule into the DOM post-render, something like layout does.

I don't know. This is getting out of hand. 😄

ramonjd commented on packages/block-editor/src/hooks/background.js:
```diff
@@ -181,8 +177,44 @@ export function BackgroundImagePanel( {
+		style: {
+			// @TODO this should be backgroundImage. How to do that?
+			// Also, maybe consider reinstating https://github.com/WordPress/gutenberg/blob/fc98542a7dbba194bb4096d49cd0bd093b63f43e/packages/block-editor/src/hooks/background.js#L82
+			background: `url( '${ encodeURI(
```
Quick question for @noisysocks or @ellatrix 

How can I update the value of a style property via `useBlockProps` or another hook? 

I want to overwrite `backgroundImage`, but I'm guessing the `useBlockProps` won't overwrite already-set props, only add new ones.... 🤔 

Thank you! 🙇🏻

ramonjd replied:
Oh never mind 😆 

I finally worked it out it's the filter order:

https://github.com/WordPress/gutenberg/blob/a8714f59987b1b252aa4fe5670bd164b7be7958d/packages/block-editor/src/hooks/index.js#L50

ramonjd commented on packages/core-data/src/resolvers.js:
```diff
@@ -896,3 +896,34 @@ export const getRevision =
 		}
 	};
+
+export const getThemeFileURI =
```
This is all highly speculative and experimental. Don't worry folks! 😅 

@andrewserong is this closer to what you were thinking?

It kinda works, but it's proving difficult to use for block props, and no idea yet how to implement for global styles.

Considering generating some temp CSS rules and injecting into the DOM 🙃 

See comment: https://github.com/WordPress/gutenberg/pull/60578/files#r1584350221

andrewserong replied:
That was the general idea I had in mind, yes, thanks for exploring it! It's a challenging one, though, as it is a bit of a messy way to handle things, but I do like the idea of us having a JS equivalent to the PHP logic, and I like how you've gone for a generic approach that could be used for things other than just the background image support 👍

draganescu commented:
> the original idea was for three intended states (full urls, theme relative urls, and featured images) 

I think featured images should be _bound_ to background-url via block bindings, not a concern of the style engine? In the UI of the editor we'll have to have a distinct way to indicate this binding, unlike static files (relative or absolute) which use the normal styling UI.

> if we remove theme as an explicit prop, we could lose the ability to do other things in the future, such as if the relative path should actually refer to an uploads directory?

When would that be the case? I mean if the background is from uploads we should maybe use the full URL?

ramonjd commented:
Thanks for continuing the discussion!

> I think featured images should be bound to background-url via block bindings, not a concern of the style engine? 

To clarify, the style engine won't be doing any of the block-editor-specifics of resolving background image paths, despite the chaotic nature of this PR's commits (all experimental). 

Totally agree that it should occur at the block level. Looking at that now for this PR.

> I mean if the background is from uploads we should maybe use the full URL?

It's pretty edge case, but it might be useful to have relative paths to uploads for testing across dev/prod environments where the installations are mirrored but the domain names are different?

I dunno. 😄 

Though possible, I'd wager it'd be detrimental to portability for a theme to add absolute paths to uploaded files in template/pattern/theme.json files.

Back to what this PR is trying to do, if we assume relative paths are _relative to the theme_, I still think that makes sense for theme assets (template/pattern/theme.json files). It doesn't have to mean we can't introduce the "source" prop later.

andrewserong commented:
> I think featured images should be bound to background-url via block bindings, not a concern of the style engine? 

What would that look like in practice? I was imagining that for the background image support, it'd feel a bit more "first party" and be a canonical part of how the block support works, so it'd be exposed in a similar way as it is in the Cover block, but as part of the `backgroundImage` object instead of a separate attribute (which is what the Cover block appears to use). Here's the current UI for the cover block for quick reference:

<img width="275" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/20ad9a30-de31-48e1-ad6a-013b17268421">

> When would that be the case? I mean if the background is from uploads we should maybe use the full URL?

My wording was likely off, but my point was more about the value of a `source` property for potential future uses rather than any one specific use case. At this stage, I'm thinking more about what choices we can make that unlock future uses rather than box us in too much.

> Back to what this PR is trying to do, if we assume relative paths are relative to the theme, I still think that makes sense for theme assets (template/pattern/theme.json files). It doesn't have to mean we can't introduce the "source" prop later.

I think that's a pragmatic path forward. We still need `backgroundImage` to support an object, as the `url` prop indicates that we'll wrap the value in a `url()` for the `background-image` CSS property, but I agree that including default behaviour that can detect if it's a relative or absolute path and then prepend the theme directory is a good path forward for now. And as you mention, it doesn't preclude a more explicit way of handling it in the future.

ramonjd committed:
Resolving urls in theme.json
Feature complete, but hacky

ramonjd commented:
Thanks for sticking with me on this one, folks.

#### Status

This PR, though hacky, is feature complete:

- relative paths to theme files are resolved in the block editor for top level theme.json styles, templates, patterns
- same above for backend
- background UI control preview icon image path is also resolved
- in both block editor and frontend, Gutenberg will first attempt to resolve the image in the stylesheet directory, which could be a child theme directory if activated, and, where that's not found, it'll default to the template theme directory (the parent).

#### Exceptions

- background styles for theme.json blocks are not yet supported. This is happening in https://github.com/WordPress/gutenberg/pull/60100
- hasn't yet removed support for theme.json block style`background.backgroundImage.source = 'theme'` in the backend. See https://github.com/WordPress/gutenberg/pull/60578#issuecomment-2087669783 Will be done in a separate PR


#### Path forward for WordPress 6.6 🤞🏻 

1. Remove requirement for `background.backgroundImage.source = 'theme'`
2. Support background image relative paths for theme.json top-level background image paths only

#### Beyond WordPress 6.6

1. Merge https://github.com/WordPress/gutenberg/pull/60100
2. Propose addition of new core data selector `getThemeFileURI`
3. Support relative paths for theme.json block style image paths
4. Support relative paths for theme template and pattern files (requires changes to background hook/block bindings, see [useBlockProps](https://github.com/WordPress/gutenberg/pull/60578/files#diff-937b387029f33f6ee39dd1594920fb292a34fec5f4ae36daa4ea89f7f061cbedR180) in this PR)

ramonjd commented:
Closing this PR as it'll be used as a reference for future PRs

See: https://github.com/WordPress/gutenberg/pull/60578#issuecomment-2087925879

ramonjd commented:
Forced push and didn't realised this PR had been closed 🙃 

So moving experiment to 

- https://github.com/WordPress/gutenberg/pull/61401
